### ISSUE TITLE
F: Including momentum and more advanced ways to update the learning rate

### ISSUE THREAD

**Issue by SigurdRiseth** [SigurdRiseth/FYS-STK4155-P1#6]
We keep our focus on OLS and Ridge regression and update our gradient descent code by including **momentum**, **AdaGrad**, **RMSprop** and **Adam** as methods for iteratively updating the learning rate (Sections 4.6 and 4.8--4.11 of the lecture notes; the lectures of weeks 37 and 38 contain several examples on how to implement these methods).  You are free to use either the analytical or the automatically differentiated gradient here -- the update rules do not care where the gradient comes from.

Discuss the results and compare the different methods applied to the one-dimensional Runge function, for example in terms of the number of iterations needed to reach a given accuracy relative to the closed-form solution, and in terms of their sensitivity to the choice of the (initial) learning rate.
